### PR TITLE
feat[ai]: J.A.Y.D.O.N music generator (#200)

### PR DESCRIPTION
**Pull request by XeIris**
Closes #200 — **J**ust **A**udio **Y**ou **D**ynamically **O**rchestrate **N**otes.

## What

The AI personas can now compose real, playable music (≤30s WAV, full General MIDI instrument set) delivered as a Discord attachment, using the skill pattern: the model reads a guide tool first, then submits a composition JSON.

**Flow:** a tiny system-prompt note advertises the capability → model calls `get_music_guide` (returns `data/skills/music-composition.md`: the JSON DSL, all 128 GM instrument names + 47 drum names, composition tips — kept out of the system prompt so it costs nothing on ordinary turns) → model calls `generate_music` with the composition as a JSON string → `utils/musicGen.ts` validates, renders offline, and attaches the WAV to the reply webhook with a `-# 🎵 composed music` prefix.

## Key pieces

- **Synth:** [`spessasynth_core`](https://github.com/spessasus/spessasynth_core) — pure-TypeScript SoundFont synth, Apache-2.0, zero deps, no WASM/native libs. Measured: ~0.3s render for 9s of 4-track audio; soundfont parses in 30ms and the parsed bank is cached (fresh processor per render).
- **Soundfont:** GeneralUser GS (~31MB, 261 presets, permissive) — **not in git**. `bun run fetch:soundfont` downloads it pinned to a commit with SHA-256 verification; the Dockerfile builder stage runs the same script. `MUSIC_SOUNDFONT_PATH` overrides.
- **Guide gating:** `generate_music` is rejected until `get_music_guide` succeeded in a *prior* tool-loop iteration, so compositions are always written with the format in context.
- **Token budget:** once the guide is read, subsequent turns in that loop get `max_tokens: 32768` (reasoning models blow the 8192 cap emitting composition JSON — found in a live model bake-off); all other turns keep 8192.
- **Validation (fail-closed):** strict whitelists and clamps — instrument names → GM programs, pitch as note-name/0–127, tempo 40–250, ≤30s at chosen tempo, ≤12 tracks, ≤1500 notes; error strings are actionable so the model self-corrects within the loop.
- **Quota:** new `MusicGenLog` table mirroring `ImageGenLog`'s atomic reserve/release transaction — 10/user/24h, failures refunded, max 2 concurrent renders, fail closed on DB errors.
- **Output:** peak-normalized 44.1kHz 16-bit stereo WAV, ≤ ~5.6MB (under Discord's cap). Audio rides the existing generated-images attachment path (webhook `files:` is type-agnostic).
- Wired for **both OpenRouter and Gemini** providers.

## Tested

- 16 new unit tests (pitch parsing, GM map completeness, validator edge cases incl. NaN smuggling, caps, drum-channel assignment); full suite 288/288, lint + typecheck clean.
- End-to-end renders verified with `afinfo` (valid WAV structure, correct duration).
- Live bake-off across the configured OpenRouter models (MiMo, Deepseek, Nemotron:free, openrouter/free) with identical prompts × 3 genres: every model that returned JSON either passed validation or self-corrected after one error-feedback retry — nothing crashed the renderer. Deepseek composed best (full 30s, 148-note orchestral); MiMo was fast but wrote short fragments, which drove the guide's "use the time budget" tip.

## Reviewer notes

- No timeout on chat completions in the tool loop: a Deepseek music turn can take ~1–3 min (reasoning). It completes; a placeholder message is a possible follow-up.
- Memory: parsed soundfont adds ~120MB RSS on first music request (lazy-loaded, cached) — fine against the 1GB container limit.
- The GPT persona (`nvidia/nemotron-3-ultra-550b-a55b:free`) timed out on every bake-off attempt (free-tier queueing) — music via that persona will likely hang until the model is swapped; accepted as-is since the persona is barely used.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added music generation support with a new composition guide and the ability to create short WAV audio tracks from structured JSON.
  * Introduced per-user daily quota tracking and integrated music generation into existing prompt/reply flows.
  * Soundfont assets are now bundled into the final build via checksum-verified fetching.
* **Bug Fixes**
  * Improved reply composition/labeling for music and attachment-only scenarios, including more accurate “search vs media” handling.
* **Tests**
  * Added coverage for pitch parsing, GM instrument/drum mapping, and strict composition validation rules.

### CODE REVIEW

File: utils/ai.ts [hunk 1]
```diff
@@ -481,9 +537,14 @@ ${systemPrompt || ''}
...
     }
+    if (musicGen && !isImageModel) {
+      if (geminiTools.length === 0) geminiTools = [{ functionDeclarations: [] }];
+      geminiTools[0].functionDeclarations.push(...musicGeminiDecls());
+    }
     const imageGenNote = !isImageModel ? buildImageGenNote(imageGen) : '';
+    const musicGenNote = !isImageModel ? buildMusicGenNote(musicGen) : '';
     const useTools = geminiTools.length > 0;
-    const toolNote = searchToolNote + imageGenNote;
+    const toolNote = searchToolNote + imageGenNote + musicGenNote;
```
[the remark this answers is not in the document]
Comment (XeIris) [reply utils/ai.ts:564]: Done in b0d4d3f — music-enabled Gemini sessions now set `generationConfig.maxOutputTokens: 32768` (the SDK only takes generationConfig per model/chat session, so the raised cap applies session-wide rather than per-turn).

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: utils/musicGen.ts [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+        };
+      }
+      channel = nextMelodicChannel;
+      nextMelodicChannel += 1;
+      if (nextMelodicChannel === DRUM_CHANNEL) nextMelodicChannel += 1;
+      if (channel > 15) {
+        return { ok: false, error: 'Error: too many melodic tracks — at most 15 (plus one "drums" track).' };
+      }
+      programs.push({ channel, program });
+    }
```
[the remark this answers is not in the document]
Comment (XeIris) [reply utils/musicGen.ts:296]: Done in b0d4d3f — `parseComposition` now rejects a second "drums" track with an actionable error, matching the guide; regression test added.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: utils/musicGen.ts [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+  while (filled < totalSamples) {
+    const now = filled / SAMPLE_RATE;
+    while (evIdx < timeline.length && timeline[evIdx].timeSec <= now) {
+      timeline[evIdx].fire(synth);
+      evIdx += 1;
+    }
+    const n = Math.min(RENDER_BLOCK, totalSamples - filled);
+    synth.process(left, right, filled, n);
+    filled += n;
+  }
```
[the remark this answers is not in the document]
Comment (XeIris) [reply utils/musicGen.ts:427]: Done in b0d4d3f — the render loop now awaits `setImmediate` every 256 blocks (~0.75s of audio ≈ ~20ms of work), so Discord events keep flowing during renders. Went with yields over a worker: a full 30s render is only ~1s of CPU, so preemption points are sufficient.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: utils/musicGen.ts [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+  let wav: Buffer | null = null;
+  inFlightRenders += 1;
+  try {
+    wav = await renderComposition(comp);
+  } catch (err) {
+    logError('[musicgen] render failed:', err);
+    await releaseQuota();
+    return { ok: false, error: 'Error: music rendering failed. Tell the user to try again later.' };
+  } finally {
+    inFlightRenders -= 1;
```
[the remark this answers is not in the document]
Comment (XeIris) [reply utils/musicGen.ts:534]: Done in b0d4d3f — the slot is claimed synchronously right after the guard (no await in between) and released in a `finally` wrapping the whole reservation+render flow.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
